### PR TITLE
🩹 Add http2 transport configuration to work around http2 upstream bug on tcp drop

### PR DESCRIPTION
**Pull request by serbrech**
**TL;DR:** 

When a TCP connection is dropped, there is a bug in the http2 implementation in go that misses to mark the connection as failed. as a result, the connection stays in the pool, and the calls using that connection hit a context deadline.
The connection can stay in the pool fur up to 17 min (with default linux kernel settings).

This configures the transport to enable ReadIdleTimeout and Ping Timeout.
internaslly, it adds the `h2` middleware on the transport which will issue http2 ping frames on the connections when no reads are occuring. if the server does not answer the ping (for example, the tcp connection was dropped), the connection is effectively closed, working around the golang upstream issue.
on the next call, a connection will be re-opened, and no context.Timeout is experienced at the app level.

**for more details, see :**
- azure sdk issue: https://github.com/Azure/azure-sdk-for-go/issues/21346#issuecomment-1699665586
- go upstream issue: https://github.com/golang/go/issues/59690

**other context:** 

We have experienced sporadic context deadline on cosmosdb operations that we attributed to server side issues.
I think this might be the actual root cause, as Julien Strohecker experienced this issue using cosmosdb sdk track 2 in a use case with higher scale than ours.

This same fix is applied in our service already.

### CODE REVIEW

File: pkg/middleware/default_http_client.go [hunk 1]
```diff
@@ -23,11 +23,16 @@ import (
...
 	"go.opentelemetry.io/otel/propagation"
+	"golang.org/x/net/http2"
 )
 
 var (
+	// defaultHTTPClient is configured with the defaultRoundTripper
 	defaultHTTPClient *http.Client
-	defaultTransport  http.RoundTripper
+	// defaultTransport is a pre-configured *http.Transport for http/2
+	defaultTransport *http.Transport
```
Comment (serbrech) [inline pkg/middleware/default_http_client.go:33]: splitting roundTripper from transport to allow testing the defaultTransport is configured, and for clarity

File: pkg/middleware/default_http_client.go [hunk 2]
```diff
@@ -37,30 +42,63 @@ func DefaultHTTPClient() *http.Client {
...
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		},
+	}
```
Comment (serbrech) [inline pkg/middleware/default_http_client.go:60]: could move this directly in the var declaration as var are instanciated before init.
or can keep it in init with the rest.

I went with keeping it in init()

**Review by haitch — APPROVED** [review]

**Review by Bryce-Soghigian — APPROVED** [review]

**Review by NickKeller — APPROVED** [review]

**Review by Bryce-Soghigian — COMMENTED** [review]
LGTM 🚀🚀🚀🚀🚀

Thanks for contributing lets make this a good holiday project
